### PR TITLE
HLRC: Get Deprecation Info API

### PR DESCRIPTION
**Pull request by hub-cap**
This commit adds the Get Deprecation Info API and associated
documentation.

Relates #29827

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MigrationRequestConverters.java [hunk 1]
```diff
@@ -48,6 +49,15 @@ static Request submitMigrateTask(IndexUpgradeRequest indexUpgradeRequest) {
         return prepareMigrateRequest(indexUpgradeRequest, false);
     }
 
+    static Request getDeprecationInfo(DeprecationInfoRequest deprecationInfoRequest) {
+        String endpoint = new RequestConverters.EndpointBuilder()
+            .addCommaSeparatedPathParts(deprecationInfoRequest.getIndices())
+            .addPathPartAsIs("_xpack", "migration", "deprecations")
```
Comment (gwbrown) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/MigrationRequestConverters.java]: In 7.0 this should use `_migration/deprecations` as of https://github.com/elastic/elasticsearch/pull/35976, although in 6.x it will have to use `_xpack`.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/migration/DeprecationInfoResponse.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        static {
+            PARSER.declareString(ConstructingObjectParser.constructorArg(), LEVEL);
+            PARSER.declareString(ConstructingObjectParser.constructorArg(), MESSAGE);
+            PARSER.declareString(ConstructingObjectParser.constructorArg(), URL);
+            PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), DETAILS);
+        }
+
+        public enum Level {
+            NONE,
+            INFO,
```
Comment (gwbrown) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/migration/DeprecationInfoResponse.java]: Soon (likely tomorrow) I'll have a PR removing `NONE` and `INFO` from this list server-side - whichever merges last will have to remove them from this list too.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MigrationIT.java [hunk 3]
```diff
@@ -73,6 +77,16 @@ public void testUpgradeWithTaskApi() throws IOException, InterruptedException {
         awaitBusy(hasUpgradeCompleted);
     }
 
+    public void testGetDeprecationInfo() throws IOException {
+        createIndex("test", Settings.EMPTY);
+        DeprecationInfoRequest request = new DeprecationInfoRequest(Collections.singletonList("test"));
+        DeprecationInfoResponse response = highLevelClient().migration().getDeprecationInfo(request, RequestOptions.DEFAULT);
+        // a test like this cannot test actual deprecations
```
Comment (gwbrown) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/MigrationIT.java]: This would be pretty easy to test in 6.x, but not so much in master. It might be worth adding a test case to 6.x, but I'm not sure.

File: docs/java-rest/high-level/migration/get-deprecation-info.asciidoc [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+include-tagged::{doc-tests-file}[{api}-request]
+--------------------------------------------------
+<1> Create a new request instance
+
+include::../execution.asciidoc[]
+
+[id="{upid}-{api}-response"]
+==== Get Deprecation Info Response
+
+The returned +{response}+ contains the cluster, index and node level deprecations.
```
Comment (gwbrown) [inline docs/java-rest/high-level/migration/get-deprecation-info.asciidoc]: This is awkward and not very clear, do you think we could reword it to something like:

```suggestion
The returned +{response}+ contains information about deprecated features currently in use at the cluster, node, and index level.
```

**Review by AthenaEryma — CHANGES_REQUESTED** [review]
Left a few comments as I've been hip-deep in the server side of this code for the last few days.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MigrationRequestConverters.java (hunk 1 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/MigrationRequestConverters.java]: Do u think its worth me doing that in a new PR after this so i can backport it back, similar to the way we did the other PRs?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/migration/DeprecationInfoResponse.java (hunk 2 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/migration/DeprecationInfoResponse.java]: ++ pls link the PR when u have it

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MigrationIT.java (hunk 3 above)
Comment (hub-cap) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/MigrationIT.java]: my fallback is that the response tests also test the response itself.

File: docs/java-rest/high-level/migration/get-deprecation-info.asciidoc (hunk 4 above)
Comment (hub-cap) [reply docs/java-rest/high-level/migration/get-deprecation-info.asciidoc]: woah what is this black magic commit suggestion thing? I kinda want to click it just to have played around with it!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MigrationRequestConverters.java (hunk 1 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/MigrationRequestConverters.java]: lol i guess i dont have a choice! 
```
   > Warnings: [[GET /{index}/_xpack/migration/deprecations] is deprecated! Use [GET /{index}/_migration/deprecations] instead.]
```

File: client/rest-high-level/src/main/java/org/elasticsearch/client/migration/DeprecationInfoResponse.java (hunk 2 above)
Comment (gwbrown) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/migration/DeprecationInfoResponse.java]: Aforementioned PR: https://github.com/elastic/elasticsearch/pull/36326

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MigrationIT.java (hunk 3 above)
Comment (gwbrown) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/MigrationIT.java]: I'll buy that argument.

**Review by AthenaEryma — APPROVED** [review]
LGTM

File: client/rest-high-level/src/main/java/org/elasticsearch/client/migration/DeprecationInfoResponse.java (hunk 2 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/migration/DeprecationInfoResponse.java]: looks like i won ;)
